### ISSUE TITLE
Add support for TypedDict __extra_items__ (PEP 728)

### ISSUE THREAD

**Issue by mharding-hpe** [python/mypy#18176]
**Feature**

[PEP 728](https://peps.python.org/pep-0728/) proposes changes to `TypedDict`, including the ability to specify the types of unspecified keys.

This has been [implemented in `typing_extensions` with the `__extra_items__` keyword](https://typing-extensions.readthedocs.io/en/latest/index.html#typing_extensions.TypedDict.__extra_items__) since version 4.10.

I would like to see support for this added to `mypy`.

**Pitch**

I am working on using mypy to validate Python code associated with rest API services. It is not uncommon for the OpenAPI specs we use to include objects whose `additionalProperties` values map to schemas. This is how OpenAPI handles the case of objects whose keywords are not known in advance, but whose values must be of a particular type.

I want to be able to type annotate these schemas in Python, but currently there is no good mechanism for doing this. That is one of the problems that PEP 728 attempts to solve.

Currently with mypy I have found a few ways to handle this (in the absence of adding the requested support):

1. As noted in the PEP, one can always suppress the `typeddict-unknown-key` errors. But of course that means no type checking is happening for the additional keys being used.
2. I haven't coded it up yet, but I suspect with overloading I could probably come up with something that works for this. Basically, I could overload the get item method of the dict, specifying the correct return types for the various pre-defined keys (as Literals), and then add a "generic" fallback get item method that returns the "extra key type".
3. It's possible I could come up with some Frankensteining of TypedDicts and defaultdicts that accomplishes this.

For both #2 and #3, it would definitely be less clean and much lengthier than using the solution in `typing_extensions`.

From my research when trying to figure out how to handle this problem, I get the impression I am not the only one trying to handle this case. That is also indicated by the fact that the current JSON schema and OpenAPI spec includes this ability. And of course, that is part of the motivation for the PEP being opened in the first place.

**Comment by layday** [comment]
The PEP has been updated and no longer specifies the `__extra_items__` *property*.  `typing_extensions` does not yet support the new parametrised form.

**Comment by mharding-hpe** [comment]
Good catch -- I missed that. In that case, I think it makes sense to wait for `typing_extensions` to catch up with the PEP before supporting it in mypy.

**Comment by mharding-hpe** [comment]
The `typing_extensions` work just got completed for this

**Comment by JukkaL** [comment]
I think all the prerequisites are ready to start work on the implementation -- we just need somebody who has time and interest to work on this.

**Comment by peacefulotter** [comment]
Any news on this? Thanks !
